CurationSorter

ferchaure opened the pull request:
Hi, guys. A working draft of the CurationSorter as we discussed it with @alejoe91. Internally it uses MergeSorting  and SplitSortingUnit (I know the naming convention is inconsistent, when we decide the names together I'll rename to the final names).

About MergeSorting :

It creates a new sortigextractor with the units combined and removes duplicates times (I had to add a function to that, probably we could move it elsewhere). I added a cache optional argument, just because checking for duplication and sort each time feels like a waste of time.

About SplitSortingUnit :

Given a list of boolean arrays (one per segment in the recording) split a unit in two parts (one for the True elements and other for the False elements). Maybe the input could be arrays with the new unit id (or index for the new ids). That will be slightly slower for one split but will make easier split a unit into multiple ones. In my mind, this could be the case for a unit you just want to remove and re assign via template matching to the other ones

About CurationSorter :

A class that internally keep a record of each stage of the sorting. It can merge, split, remove and rename units. It can undo and redo changes and keep a graph with the changes applied. An example of how to use it [here](https://gist.github.com/ferchaure/fd6ae54c3258b8b65b21fffb9b9f48f6)

At the begging, I want it to behave like a sortingextractor, but  It doesn't inheritance from basesorter, because its properties like units_id would change after each operation.  For that reason any method/attribute call not implemented will try to use the one on the current sorting.  Probably, that's not necessary and using the property `.sorting` to refer to the current sorting is good enough.

I didn't implement a save method, then if the user saves it, it will save just the current sorting. Probably I will need a way to save the graph with the history as well.

**To-Do**
- Maybe move these classes to other folder 
- Tests
- Documentation
- Save the graph with the details of the curation.
- Include comments in the curation, like number of spikes, or proportion of the spikes split to the new unit

alejoe91 commented on spikeinterface/core/curationsorting.py:
```diff
@@ -0,0 +1,119 @@
+from .mergesorting import MergeSorting
+from .splitsortingunit import SplitSortingUnit
+import networkx as nx
```
I think that the graph should be optional, so we don't enforce networkx installation. 

Note that `networkx` is installed in `full` installation, so most of the times it will be available, but if one only wants to use the core it should not be required.

alejoe91 commented on spikeinterface/core/__init__.py:
```diff
@@ -28,6 +28,11 @@
 from .unitsaggregationsorting import UnitsAggregationSorting, aggregate_units
 
+# utility sorters
+from .mergesorting import MergeSorting
+from .splitsortingunit import SplitSortingUnit
```
```suggestion
from .mergesorting import MergeUnitSorting
from .splitsortingunit import SplitUnitSorting
```

IMO we can use these names for consistency. What do you think?

ferchaure replied:
maybe... MergeSortingUnit and SplitSortingUnit, sound clearest for me. Or even something like SplitUnit and MergeUnits. All of them works are better than the current ones (I started playing with graphs and went lazy with the details). Perhaps @samuelgarcia has a strong preference for any of them?

alejoe91 commented on spikeinterface/core/curationsorting.py:
```diff
@@ -0,0 +1,119 @@
+
+    """
+
+    def __init__(self, parent_sorting):
```
We could inherit from `BaseSorting` too. I think it would make it convenient so that all functions would be directly available on the top level. All we need to do is instantiate the object with the initial sorting unit ids, and then override `get_unit_ids`, `get_unit_spike_train`, and `get/set_property`. Or (but it needs to be tested), we can do something like: 
```
self = self._sorting_stages[self._sorting_stages_i]
```
and we have to call this whenever an action is made. Let's discuss about this!

ferchaure replied:
the issue with inherit from BaseSorting is that we will need to keep updating the functions if we change something in the BaseSorting 

In a way curationsorting is not a proper BaseSorting, but a basesorting handler. Using the methods as a BaseSorting  is just for simplicity, I'm not even sure if using the `sorting` property to access the current sorting is enough

alejoe91 commented on spikeinterface/core/curationsorting.py:
```diff
@@ -0,0 +1,119 @@
+        edges = edges + [(node_t(unit2split,i), node_t(u,i+1)) for u in new_units_ids]
+        self._add_new_stage(new_sorting, edges)
+
+    def merge(self,units2merge, rm_dup_delta_time=0.4, cache=False):
```
```suggestion
    def merge(self,units2merge, delta_time=0.4, cache=False):
```
and explain in docstring

alejoe91 commented on spikeinterface/core/curationsorting.py:
```diff
@@ -0,0 +1,119 @@
+        nodes = [n.id for n in self._graphs[self._sorting_stages_i].nodes()]
+        return max(nodes)+1
+
+    def split(self,unit2split, indices_list):
```
Should `new_unit_ids` be an optional arg as well? Note that unit ids could also be strings!

ferchaure replied:
Ok, I will check for strings as units. Probably an int cast at the begging 

The comment position is wrong. You said something like by default the split units id will be the next available units id?

alejoe91 commented on spikeinterface/core/curationsorting.py:
```diff
@@ -0,0 +1,119 @@
+    def sorting(self):
+        return self._sorting_stages[self._sorting_stages_i]
+
+    def __getattr__(self,name):
+        #any method not define for this class will try to use the current 
+        #  sorting stage. In that whay this class will behave as a sortingextractor
+        current_sorting = self._sorting_stages[self._sorting_stages_i]
+
+        attr = object.__getattribute__(current_sorting, name)
+        return attr
```
Nice!!!

alejoe91 commented on spikeinterface/core/mergesorting.py:
```diff
@@ -0,0 +1,150 @@
+
+from .basesorting import BaseSorting, BaseSortingSegment
+
+class MergeSorting(BaseSorting):
```
```suggestion
class MergeUnitSorting(BaseSorting):
```

alejoe91 commented on spikeinterface/core/mergesorting.py:
```diff
@@ -0,0 +1,150 @@
+        Sorting object with the selected units merged
+    """
+
+    def __init__(self, parent_sorting, units2merge, merge_unit_id=None, rm_dup_delta_time=0.4, cache=False):
```
```suggestion
    def __init__(self, parent_sorting, units_to_merge, merge_unit_id=None, delta_time=0.4, cache=False):
```

alejoe91 commented on spikeinterface/core/mergesorting.py:
```diff
@@ -0,0 +1,150 @@
+    units2merge: Sorting
+        The sorting object
+    merge_unit_id: list
+        The folder where waveforms are cached
```
?

ferchaure replied:
copy/paste the doc format from somewhere :facepalm:

alejoe91 commented on spikeinterface/core/splitsortingunit.py:
```diff
@@ -0,0 +1,103 @@
+from copy import deepcopy
+from .basesorting import BaseSorting, BaseSortingSegment
+
+class SplitSortingUnit(BaseSorting):
```
```suggestion
class SplitUnitSorting(BaseSorting):
```

alejoe91 reviewed (changes requested):
Thank you @ferchaure 

A few comments! Also can you add `networkx` in the `requirements_test.txt`?

ferchaure commented:
I already did some of the changes (the resolved ones) and added support for ids with strings, check the updated gist

ferchaure commented:
Hi guys, sorry for the late update. I implemented the changes we talked about. Let me know what do you think

alejoe91 reviewed (changes requested):
Awesome Fer!! Could you add some small tests?

Maybe also with a multi-segment object! :)

ferchaure commented:
test added... I don't like the behavior of merging all units when the properties are different (or properties_policy='remove'). In this implementation I have to remove the property because is not possible to call `set_property `with an empty `ids`.
Ideally I would like to call:
`sorting.set_property('someprop',np.array([],dtype=float),[])`
to set as nan the property 'someprop' for all the units. 
@samuelgarcia You you think that can be be the behavior of set_property? (I can try to add that to this pull request)

alejoe91 commented:
> test added... I don't like the behavior of merging all units when the properties are different (or properties_policy='remove'). In this implementation I have to remove the property because is not possible to call `set_property `with an empty `ids`. Ideally I would like to call: `sorting.set_property('someprop',np.array([],dtype=float),[])` to set as nan the property 'someprop' for all the units. @samuelgarcia You you think that can be be the behavior of set_property? (I can try to add that to this pull request)

@ferchaure I don't understand exatly what you mean. Why should the property for all units be removed? Shouldn't we only set the properties for the merged unit as empty? In that case we can do something like this:

```
# self is the MergeUnitsSorting
self.set_property("property_name", values=self._parent_sorting.get_property("property_name")[idxs_without_merged],
                            ids=unit_ids_without_merged)
```

ferchaure commented:
The issue arises when all the units are merge in one. Then there is not a way to call set_property giving just an empty values

alejoe91 commented:
I see! That's a pretty edge use case, but in that case shouldn't this work?
```
sorting.set_property('someprop', np.array([np.nan] * len(sorting.unit_ids)))
```

ferchaure commented:
> ```
> sorting.set_property('someprop', np.array([np.nan] * len(sorting.unit_ids)))
> ```

That could change the type of the property, and could break something for the user. Maybe is not crazy to ask set_property to work with empty `ids `and empty `values`. Extrapolating the current behavior set_property will set all the units with the empty value corresponding to the type of `values`.

ferchaure commented:
Ok, I fixed that odd case.  A small change in merge and a change in  `set_property`... With this change, if you create a property with an empty ids list and empty value array (of supported type), `set_property` just set the empty value for all of them. I think it just expand the usual behavior of the method. 

On the other hand, I changed `ids_to_indices`, because when it was called with an empty ids, numpy creates an array of floats by default.  And I you tried to index something with that empty array you will get an error (indexing with floats). Now the output is always int.
